internal/controllers/configuration/kongupstreampolicy_utils.go

```diff
@@ -0,0 +1,178 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	gatewaycontroller "github.com/kong/kubernetes-ingress-controller/v3/internal/controllers/gateway"
+	"github.com/kong/kubernetes-ingress-controller/v3/internal/gatewayapi"
+	kongv1beta1 "github.com/kong/kubernetes-ingress-controller/v3/pkg/apis/configuration/v1beta1"
+)
+
+type serviceStatus struct {
+	service           corev1.Service
+	acceptedCondition metav1.Condition
+}
+
+// -----------------------------------------------------------------------------
+// KongUpstreamPolicy Controller - Reconciler Helpers
+// -----------------------------------------------------------------------------
+
+// enforceKongUpstreamPolicyStatus gets a list of services (ancestors) along with their desired status and enforce them
+// in the KongUpstreamPolicy status.
+func (r *KongUpstreamPolicyReconciler) enforceKongUpstreamPolicyStatus(ctx context.Context, oldPolicy *kongv1beta1.KongUpstreamPolicy, servicesStatus []serviceStatus) (bool, error) {
+	newPolicyStatus := gatewayapi.Policystatus{}
+	if len(servicesStatus) > 0 {
+		newPolicyStatus.Ancestors = make([]gatewayapi.PolicyAncestorStatus, 0, len(servicesStatus))
+	}
+	for _, ss := range servicesStatus {
+		newPolicyStatus.Ancestors = append(newPolicyStatus.Ancestors,
+			gatewayapi.PolicyAncestorStatus{
+				AncestorRef: gatewayapi.ParentReference{
+					Group:     lo.ToPtr(gatewayapi.Group("core")),
+					Kind:      lo.ToPtr(gatewayapi.Kind("Service")),
+					Namespace: lo.ToPtr(gatewayapi.Namespace(ss.service.Namespace)),
+					Name:      gatewayapi.ObjectName(ss.service.Name),
+				},
+				ControllerName: gatewaycontroller.GetControllerName(),
+				Conditions: []metav1.Condition{
+					ss.acceptedCondition,
+				},
+			},
+		)
+	}
+	if isStatusUpdated := isPolicyStatusUpdated(oldPolicy.Status, newPolicyStatus); !isStatusUpdated {
+		newPolicy := oldPolicy.DeepCopy()
+		newPolicy.Status = newPolicyStatus
+		return true, r.Client.Status().Patch(ctx, newPolicy, client.MergeFrom(oldPolicy))
+	}
+	return false, nil
+}
+
+// buildServicesStatus creates a list of services with their conditions associated.
+func (r *KongUpstreamPolicyReconciler) buildServicesStatus(ctx context.Context, services []corev1.Service) ([]serviceStatus, error) {
+	// prepare a service mapping to be used in subsequent operations
+	mappedServices := make(map[string]serviceStatus)
+	for _, service := range services {
+		acceptedCondition := metav1.Condition{
+			Type:               string(gatewayapi.PolicyConditionAccepted),
+			Status:             metav1.ConditionTrue,
+			Reason:             string(gatewayapi.PolicyReasonAccepted),
+			LastTransitionTime: metav1.Now(),
+		}
+		mappedServices[buildServiceReference(service.Namespace, service.Name)] = serviceStatus{
+			service:           service,
+			acceptedCondition: acceptedCondition,
+		}
+	}
+
+	for _, service := range services {
+		httpRoutes := &gatewayapi.HTTPRouteList{}
```

Comment:
Here we can also consider using reference indexer to save the reference relations.

Reply:
Yes, that's what I've done.See line 126